src/EFCore.PG/Storage/Internal/Mapping/NpgsqlArrayTypeMapping.cs

```diff
@@ -133,22 +133,25 @@ static Type MakeCollectionType(Type collectionType, Type elementType, bool isEle
             }
         }
 
-        // TODO: Confirm model vs. provider type here!
         // We do GetElementType for multidimensional arrays - these don't implement generic IEnumerable<>
-        var modelElementType = typeof(TCollection).TryGetElementType(typeof(IEnumerable<>)) ?? typeof(TCollection).GetElementType();
+        var elementType = typeof(TCollection).TryGetElementType(typeof(IEnumerable<>)) ?? typeof(TCollection).GetElementType();
 
-        Check.DebugAssert(modelElementType is not null, "modelElementType cannot be null");
+        Check.DebugAssert(elementType is not null, "modelElementType cannot be null");
 
+#pragma warning disable EF1001
         var comparer = typeof(TCollection).IsArray && typeof(TCollection).GetArrayRank() > 1
             ? null // TODO: Value comparer for multidimensional arrays
             : (ValueComparer?)Activator.CreateInstance(
-                modelElementType.IsNullableValueType()
-                    ? typeof(NullableValueTypeListComparer<>).MakeGenericType(modelElementType.UnwrapNullableType())
-                    : typeof(ListComparer<>).MakeGenericType(elementTypeMapping.Comparer.Type),
-                elementTypeMapping.Comparer);
+            elementType.IsNullableValueType()
+                ? typeof(NullableValueTypeListComparer<>).MakeGenericType(elementType.UnwrapNullableType())
+                : elementMapping.Comparer.Type.IsAssignableFrom(elementType)
+                    ? typeof(ListComparer<>).MakeGenericType(elementType)
+                    : typeof(ObjectListComparer<>).MakeGenericType(elementType),
+            elementMapping.Comparer.ToNullableComparer(elementType)!);
```

Comment:
@ajcvickers it seems that ToNullableComparer is now required (without it I got an exception). That's not a problem, but it's Internal, which could be problematic...

Am I doing the right thing here etc.? Should we make it public?